torch_frame/nn/encoder/stype_encoder.py

```diff
@@ -721,49 +722,51 @@ def __init__(
         stype: Optional[stype] = None,
         post_module: Optional[Module] = None,
         na_strategy: Optional[NAStrategy] = None,
-        in_channels: Optional[int] = None,
-        model: Optional[torch.nn.Module] = None,
+        col_to_model_cfg: Optional[Dict[str, TextModelConfig]] = None,
     ):
-        if in_channels is None:
-            raise ValueError("Please manually specify the `in_channels`, "
-                             "which is the text embedding dimensionality.")
-        if model is None:
-            raise ValueError("Please manually specify the `model`, "
-                             "which outputs embeddings that will be feed "
-                             "to the linear layer.")
+        if col_to_model_cfg is None:
+            raise ValueError("Please manually specify the "
+                             "`col_to_model_cfg`, which outputs "
+                             "embeddings that will be feed to the linear "
+                             "layer.")
 
-        self.in_channels = in_channels
+        self.col_to_model_cfg = col_to_model_cfg
 
         super().__init__(out_channels, stats_list, stype, post_module,
                          na_strategy)
-        self.model = model
 
     def init_modules(self):
         super().init_modules()
-        num_cols = len(self.stats_list)
-        self.weight = Parameter(
-            torch.empty(num_cols, self.in_channels, self.out_channels))
-        self.bias = Parameter(torch.empty(num_cols, self.out_channels))
+        self.weight = torch.nn.ParameterDict()
+        self.bias = torch.nn.ParameterDict()
+        for col_name, model_cfg in self.col_to_model_cfg.items():
+            self.weight[col_name] = Parameter(
+                torch.empty(model_cfg.out_channels, self.out_channels))
+            self.bias[col_name] = Parameter(torch.empty(self.out_channels))
         self.reset_parameters()
 
     def reset_parameters(self):
         super().reset_parameters()
-        torch.nn.init.normal_(self.weight, std=0.01)
-        torch.nn.init.zeros_(self.bias)
+        for col_name in self.weight:
+            torch.nn.init.normal_(self.weight[col_name], std=0.01)
+            torch.nn.init.zeros_(self.bias[col_name])
 
     def encode_forward(
         self,
         feat: TensorData,
         col_names: Optional[List[str]] = None,
     ) -> Tensor:
-        x = self.model(feat)
-        # [batch_size, num_cols, in_channels] *
-        # [num_cols, in_channels, out_channels]
-        # -> [batch_size, num_cols, out_channels]
-        x_lin = torch.einsum("ijk,jkl->ijl", x, self.weight)
-        # [batch_size, num_cols, out_channels] + [num_cols, out_channels]
-        # -> [batch_size, num_cols, out_channels]
-        x = x_lin + self.bias
+        xs = []
+        for i, col_name in enumerate(col_names):
+            # [batch_size, in_channels]
+            x = self.col_to_model_cfg[col_name].model(
+                {key: feat[key][:, i]
```

Comment:
I don't like this column access here. We can merge it like this now but I think eventually we want to refactor the code and store the tensors by the same encoders than by the same stypes.

Reply:
Is `col_name` really neccessary? Since the order of the columns would not change, we can just process and store `col_to_model_cfg` to a list instead of dictionary.

Reply:
How do you know the column order until you see tensor_frame's `col_names`?

Reply:
I see.